archive/analysis_metrics.py

```diff
@@ -0,0 +1,109 @@
+""" Some functions used to inspect and anlysize our data.
+"""
+
+from functools import reduce
+import json
+from math import sqrt
+from itertools import tee as copy_iterable
+from statistics import median, mean
+import sys
+from typing import Iterator
+import pandas as pd  # type: ignore
+
+
+def calculate_metrics(df: pd.DataFrame) -> tuple[float, float, float]:
+    """Calculate aggregate min, max, and avg for dataframe
+
+    Args:
+        df (pd.DataFrame): Dataframe to calculate on
+
+    Returns:
+        tuple[float, float, float]: tuple containing the avg, min and max speed
+    """
+    # per entry: max, min, avg speed across entire segment, median
+    df["avg_speed"] = df["speeds"].apply(mean)  # type: ignore
+    df["max_speed"] = df["speeds"].apply(max)
+    df["min_speed"] = df["speeds"].apply(min)
+    df["median"] = df["speeds"].apply(median)  # type: ignore
+
+    # for all entries: median_avg, median_min, median_max
+    median_avg = df["avg_speed"].median()
+    median_min = df["min_speed"].median()
+    median_max = df["max_speed"].median()
+
+    return (median_avg, median_min, median_max)
+
+
+def verify_solution(l: tuple[Iterator[bool], Iterator[bool]]):
+    """Auxiliary fuction used in filter_segments, to verify the solution works.
+    I assume that trips do not loop back and go through the same
+    segment more than once. This function is a sanitity check for this.
+
+    Args:
+        l (list[bool]): list with boolean mask of all coordinates that are in
+        the segment of interest.
+    """
+    safe_cmp = lambda b, acc: True if len(acc) == 0 else acc[-1] != b
+    reduced: Iterator[
+        bool
+    ] = reduce(  # convert [False, False, True, True, True, False] -> [False, True, False]
+        lambda acc, bool: acc + [bool] if safe_cmp(bool, acc) else acc, l, []  # type: ignore
+    )
```

Comment:
```suggestion
    cmp_if_different = lambda b, acc: True if len(acc) == 0 else acc[-1] != b
    reduced: Iterator[
        bool
    ] = reduce(  # convert [False, False, True, True, True, False] -> [False, True, False]
        lambda acc, bool: acc + [bool] if cmp_if_different(bool, acc) else acc, l, []  # type: ignore
    )
```

Synes det er lidt mere læsbart da det trippede mig lidt op at en cmp funktion lavede `!=`

Reply:
Og jeg ved godt denne pr bare er til docstrings